Adopt more smart pointers in AbortSignal

cdumez opened the pull request:
#### fd24e6ddab25326a522eeec9fdea6a8c08c340f0
<pre>
Adopt more smart pointers in AbortSignal
<a href="https://bugs.webkit.org/show_bug.cgi?id=267863">https://bugs.webkit.org/show_bug.cgi?id=267863</a>

Reviewed by Geoffrey Garen.

* Source/WebCore/Modules/fetch/FetchRequest.cpp:
(WebCore::FetchRequest::initializeWith):
(WebCore::FetchRequest::clone):
* Source/WebCore/Modules/fetch/FetchRequest.h:
* Source/WebCore/bindings/js/JSDOMGlobalObject.cpp:
(WebCore::JSC_DEFINE_HOST_FUNCTION):
* Source/WebCore/bindings/js/JSDOMWrapper.h:
(WebCore::JSDOMWrapper::protectedWrapped const):
* Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
(GenerateHeader):
* Source/WebCore/bindings/scripts/test/JS/JSDedicatedWorkerGlobalScope.h:
(WebCore::JSDedicatedWorkerGlobalScope::protectedWrapped const):
* Source/WebCore/bindings/scripts/test/JS/JSExposedStar.h:
(WebCore::JSExposedStar::protectedWrapped const):
* Source/WebCore/bindings/scripts/test/JS/JSLocalDOMWindow.h:
(WebCore::JSLocalDOMWindow::protectedWrapped const):
* Source/WebCore/bindings/scripts/test/JS/JSPaintWorkletGlobalScope.h:
(WebCore::JSPaintWorkletGlobalScope::protectedWrapped const):
* Source/WebCore/bindings/scripts/test/JS/JSServiceWorkerGlobalScope.h:
(WebCore::JSServiceWorkerGlobalScope::protectedWrapped const):
* Source/WebCore/bindings/scripts/test/JS/JSSharedWorkerGlobalScope.h:
(WebCore::JSSharedWorkerGlobalScope::protectedWrapped const):
* Source/WebCore/bindings/scripts/test/JS/JSTestDOMJIT.h:
(WebCore::JSTestDOMJIT::protectedWrapped const):
* Source/WebCore/bindings/scripts/test/JS/JSTestDefaultToJSONIndirectInheritance.h:
(WebCore::JSTestDefaultToJSONIndirectInheritance::protectedWrapped const):
* Source/WebCore/bindings/scripts/test/JS/JSTestDefaultToJSONInherit.h:
(WebCore::JSTestDefaultToJSONInherit::protectedWrapped const):
* Source/WebCore/bindings/scripts/test/JS/JSTestDefaultToJSONInheritFinal.h:
(WebCore::JSTestDefaultToJSONInheritFinal::protectedWrapped const):
* Source/WebCore/bindings/scripts/test/JS/JSTestEventConstructor.h:
(WebCore::JSTestEventConstructor::protectedWrapped const):
* Source/WebCore/bindings/scripts/test/JS/JSTestEventTarget.h:
(WebCore::JSTestEventTarget::protectedWrapped const):
* Source/WebCore/bindings/scripts/test/JS/JSTestNode.h:
* Source/WebCore/bindings/scripts/test/JS/JSTestPromiseRejectionEvent.h:
(WebCore::JSTestPromiseRejectionEvent::protectedWrapped const):
* Source/WebCore/bindings/scripts/test/JS/JSWorkerGlobalScope.h:
(WebCore::JSWorkerGlobalScope::protectedWrapped const):
* Source/WebCore/bindings/scripts/test/JS/JSWorkletGlobalScope.h:
(WebCore::JSWorkletGlobalScope::protectedWrapped const):
* Source/WebCore/dom/AbortSignal.cpp:
(WebCore::AbortSignal::timeout):
(WebCore::AbortSignal::any):
(WebCore::AbortSignal::signalAbort):
(WebCore::AbortSignal::signalFollow):

Canonical link: <a href="https://commits.webkit.org/273317@main">https://commits.webkit.org/273317@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4b86f8763edc9dd892e4dc9e82e981fe37f07a34

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/35054 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/13962 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/37152 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/37814 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/31659 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/36213 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/16348 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/11040 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/30598 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/35598 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/11834 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/31273 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/10373 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/10432 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/31368 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/39061 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/31898 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/31699 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/36440 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/10534 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/8468 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/34429 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/12325 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/11064 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4519 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/11388 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

geoffreygaren commented on Source/WebCore/bindings/js/JSDOMWrapper.h:
```diff
@@ -77,6 +77,7 @@ class JSDOMWrapper : public JSDOMObject {
     using DOMWrapped = ImplementationClass;
 
     ImplementationClass& wrapped() const { return m_wrapped; }
+    Ref<ImplementationClass> protectedWrapped() const { return m_wrapped; }
```
I guess eventually we may end up with three accessors per data member for certain data members: unprotected, protected, and weak. On the one hand, that feels a little ick. On the other hand, I think it's correct?

You could say that we should only have the protected accessor, but: If you already have a reference you don't want to waste compute on making another; and we probably can't ask the compiler to make that cost disappear; and if a class holds a weak pointer internally it can't guarantee a non-null return value.

You could say that we should only have the unprotected accessor, but: If you don't already have a reference, you don't want to have to write an extra line of code just to get one.

So I think this is how we want it to be.

geoffreygaren replied:
I guess the alternative would be for the caller to do

`ref(wrapped())->doSomething();`

or, in the general case,

`ref(thing1())->ref(thing2())->ref(thing3())->doSomething()`

But I think that's worse. Usually there are more callers than callees, so it's nice to make the callee a little more verbose so that the caller can be a little less verbose.

cdumez replied:
> I guess eventually we may end up with three accessors per data member for certain data members: unprotected, protected, and weak.

Can you come up with an example where we'd ever need a weak getter? I don't think this is currently a thing in our code base and I can't come up with a good use case. It's either `unprotected + protected` (if refcounted) or `unprotected + checked` if not refcounted.

cdumez replied:
We want the unprotected one for chaining and the protected one for the last call in the chain:
```
foo().protectedBar()->doWork();
```

cdumez replied:
Note that none of this is new to this patch. Those are well established patterns since Ryosuke started smart pointer adoption. If you are not a fan of the current pattern, I suggest with start a Slack discussion with Ryosuke as he came up with this design and there is a lot of code impacted by it.

geoffreygaren replied:
I am saying that I am a fan of the pattern, but it took me some thinking to get there.

geoffreygaren replied:
A possible example of a weak pointer accessor is:

AccessibilityImageMapLink has a private data member

`    WeakPtr<HTMLAreaElement, WeakPtrImplWithEventTargetData> m_areaElement;
`
and a public accessor

`    HTMLAreaElement* areaElement() const { return m_areaElement.get(); }
`

Now that I think of it, I guess there's no case where the caller needs a WeakPtr accessor here. The caller can never chain the accessor into another call since they need to check for nullptr first. And if the caller wants to assign to their own WeakPtr, they can do so. 

A possible extreme example is if we have a hot loop that constructs WeakPtrs. In such a loop we might want a WeakPtr accessor because the WeakPtr copy constructor is slightly faster than the WeakPtr pointer constructor. I can't think of a case like that, though.

geoffreygaren reviewed (approved):
r=me

cdumez committed:
Adopt more smart pointers in AbortSignal
https://bugs.webkit.org/show_bug.cgi?id=267863

Reviewed by Geoffrey Garen.

* Source/WebCore/Modules/fetch/FetchRequest.cpp:
(WebCore::FetchRequest::initializeWith):
(WebCore::FetchRequest::clone):
* Source/WebCore/Modules/fetch/FetchRequest.h:
* Source/WebCore/bindings/js/JSDOMGlobalObject.cpp:
(WebCore::JSC_DEFINE_HOST_FUNCTION):
* Source/WebCore/bindings/js/JSDOMWrapper.h:
(WebCore::JSDOMWrapper::protectedWrapped const):
* Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
(GenerateHeader):
* Source/WebCore/bindings/scripts/test/JS/JSDedicatedWorkerGlobalScope.h:
(WebCore::JSDedicatedWorkerGlobalScope::protectedWrapped const):
* Source/WebCore/bindings/scripts/test/JS/JSExposedStar.h:
(WebCore::JSExposedStar::protectedWrapped const):
* Source/WebCore/bindings/scripts/test/JS/JSLocalDOMWindow.h:
(WebCore::JSLocalDOMWindow::protectedWrapped const):
* Source/WebCore/bindings/scripts/test/JS/JSPaintWorkletGlobalScope.h:
(WebCore::JSPaintWorkletGlobalScope::protectedWrapped const):
* Source/WebCore/bindings/scripts/test/JS/JSServiceWorkerGlobalScope.h:
(WebCore::JSServiceWorkerGlobalScope::protectedWrapped const):
* Source/WebCore/bindings/scripts/test/JS/JSSharedWorkerGlobalScope.h:
(WebCore::JSSharedWorkerGlobalScope::protectedWrapped const):
* Source/WebCore/bindings/scripts/test/JS/JSTestDOMJIT.h:
(WebCore::JSTestDOMJIT::protectedWrapped const):
* Source/WebCore/bindings/scripts/test/JS/JSTestDefaultToJSONIndirectInheritance.h:
(WebCore::JSTestDefaultToJSONIndirectInheritance::protectedWrapped const):
* Source/WebCore/bindings/scripts/test/JS/JSTestDefaultToJSONInherit.h:
(WebCore::JSTestDefaultToJSONInherit::protectedWrapped const):
* Source/WebCore/bindings/scripts/test/JS/JSTestDefaultToJSONInheritFinal.h:
(WebCore::JSTestDefaultToJSONInheritFinal::protectedWrapped const):
* Source/WebCore/bindings/scripts/test/JS/JSTestEventConstructor.h:
(WebCore::JSTestEventConstructor::protectedWrapped const):
* Source/WebCore/bindings/scripts/test/JS/JSTestEventTarget.h:
(WebCore::JSTestEventTarget::protectedWrapped const):
* Source/WebCore/bindings/scripts/test/JS/JSTestNode.h:
* Source/WebCore/bindings/scripts/test/JS/JSTestPromiseRejectionEvent.h:
(WebCore::JSTestPromiseRejectionEvent::protectedWrapped const):
* Source/WebCore/bindings/scripts/test/JS/JSWorkerGlobalScope.h:
(WebCore::JSWorkerGlobalScope::protectedWrapped const):
* Source/WebCore/bindings/scripts/test/JS/JSWorkletGlobalScope.h:
(WebCore::JSWorkletGlobalScope::protectedWrapped const):
* Source/WebCore/dom/AbortSignal.cpp:
(WebCore::AbortSignal::timeout):
(WebCore::AbortSignal::any):
(WebCore::AbortSignal::signalAbort):
(WebCore::AbortSignal::signalFollow):

Canonical link: https://commits.webkit.org/273317@main
